java/server/src/org/openqa/grid/web/servlet/NodeSessionsServlet.java

```diff
@@ -0,0 +1,154 @@
+import org.openqa.selenium.remote.http.HttpRequest;
+import org.openqa.selenium.remote.http.HttpResponse;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.io.Writer;
+import java.net.URL;
+import java.nio.charset.Charset;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * API to query all the sessions that are currently running in the hub.
+ */
+public class NodeSessionsServlet extends RegistryBasedServlet {
+
+  private final Json json = new Json();
+
+  public NodeSessionsServlet() {
+    this(null);
+  }
+
+  public NodeSessionsServlet(GridRegistry registry) {
+    super(registry);
+  }
+
+  @Override
+  protected void doGet(HttpServletRequest req, HttpServletResponse rsp) throws IOException {
+    process(rsp);
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException {
+    if (req.getInputStream() != null) {
+      process(resp);
+    } else {
+      process(resp);
+    }
+  }
+
+  protected void process(HttpServletResponse response) throws IOException {
+    response.setContentType("application/json");
+    response.setCharacterEncoding("UTF-8");
+    response.setStatus(200);
+    Map<String, Object> proxies = new TreeMap<>();
+    try (Writer writer = response.getWriter(); JsonOutput out = json.newOutput(writer)) {
+      proxies.put("success", true);
+      proxies.put("proxies", extractSessionsFromAllProxies());
+      out.write(proxies);
+    }
+  }
+
+  private List<Map<String, Object>> extractSessionsFromAllProxies() {
+    List<Map<String, Object>> results = new LinkedList<>();
+    List<RemoteProxy> proxies = getRegistry().getAllProxies().getBusyProxies();
+    for (RemoteProxy proxy : proxies) {
+      Map<String, Object> sessionsInProxy = new TreeMap<>(extractSessionInfo(proxy));
+      if (sessionsInProxy.isEmpty()) {
+        continue;
+      }
+      Map<String, Object> res = new TreeMap<>();
+      res.put("proxyId", proxy.getId());
+      res.put("proxyRemoteHost", proxy.getRemoteHost().toString());
+      res.put("sessions", sessionsInProxy);
+      results.add(res);
+    }
+    return results;
+  }
+
+  private static Map<String, Object> extractSessionInfo(RemoteProxy proxy) {
+    try {
+      URL url = proxy.getRemoteHost();
+      HttpRequest req = new HttpRequest(HttpMethod.GET, "/wd/hub/sessions");
+      HttpResponse rsp = proxy.getHttpClient(url).execute(req);
+      try (InputStream in = new ByteArrayInputStream(asBytes(rsp));
+           Reader reader = new InputStreamReader(in, getContentEncoding(rsp))) {
+        Object body = new Json().newInput(reader).read(Object.class);
+        if (body instanceof Map) {
```

Comment:
Instead of doing this, you can get the json coercer to do all the hard work:

```
Map<String, Object> body = json.newInput(reader).read(Json.MAP_TYPE);
```

Also, you're going to leak the `JsonInput`: best open that in the try statement above.